Setup GUI + init rework

jaj22 opened the pull request:
### What type of PR is this.
1. [ ] Bug
2. [ ] Change
3. [X] Enhancement

### What have you changed and why?
This PR adds a setup GUI to select and manipulate saved games, select factions and change parameters. This required a major init rework, with some overdue refactors carried out along the way.

Functionality is largely complete, but I need to fill out a lot of doc headers and do more faction-specific equipment testing. Backing up your profile vars files is strongly recommended although I didn't break them for a while.

### Please specify which Issue this PR Resolves.
closes #XXXX

### Please verify the following and ensure all checks are completed.
1. [X] Have you loaded the mission in LAN host?
2. [X] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [ ] No
2. [X] Yes (Please provide further detail below.)

Stuff not adequately tested yet:
- Loot list generation.
- Any factions except vanilla, RHS and 3CBF.
- Campaign completion flagging.
- Probably other stuff I forgot.

killerswin2 commented on A3A/addons/core/Templates/fn_detector.sqf:
```diff
@@ -76,12 +76,6 @@ if (isClass (configfile >> "CfgPatches" >> "Ivory_Data")) then {A3A_hasIvory = t
-//ADV-CPR Pike Edition detection
-if (A3A_hasACEMedical && isClass (configFile >> "CfgPatches" >> "adv_aceCPR")) then {A3A_hasADV = true; Info("ADV Detected.") };
-
-//KAT medical detection
```
I'm confused on why you removed this. Was it replaced some where that I didn't see?

jaj22 replied:
Yeah, it's moved back a step to initVarCommon. Detector and selector aren't currently called, just left in for reference. I'll probably kill them before release.

killerswin2 commented on A3A/addons/core/Templates/Templates.hpp:
```diff
@@ -1,533 +1,596 @@
+        flagTexture = "a3\data_f\flags\flag_fia_co.paa";
+        name = "CUP civs";
+        file = "CUP_Civ";
     };
```
UNS got removed somehow

jaj22 replied:
Unsung now added and working, even found a logo.

killerswin2 commented on A3A/addons/core/functions/init/fn_initVarCommon.sqf:
```diff
@@ -108,9 +90,25 @@ if (A3A_hasTFARBeta) then {A3A_hasTFAR = false};
+
+// Here because it changes the verification logic apparently
+// should probably detect CBA instead, or see how well the 2.10 compatibleItems works
+A3A_hasCUP = isClass (configFile >> "cfgPatches" >> "CUP_BaseConfigs");
```
the verfication is approved Bob just needs to merge it in. You can remove this if you want when he merges in the update.

jaj22 replied:
Yeah, the more of these I can kill, the better. One day I'll purge all the hasIFA references.

Bob-Murphy replied:
#2466 is merged

killerswin2 commented on A3A/addons/core/Templates/fn_detector.sqf:
```diff
@@ -76,12 +76,6 @@ if (isClass (configfile >> "CfgPatches" >> "Ivory_Data")) then {A3A_hasIvory = t
 //TCGM_BikeBackpack Detection
```
honestly I would just kill this file if we aren't going to use it.

killerswin2 commented on A3A/addons/core/functions/Base/fn_aggressionUpdateLoop.sqf:
```diff
@@ -40,7 +40,7 @@ while {true} do
     private _lastScale = A3A_balancePlayerScale;
     A3A_balancePlayerScale = (A3A_activePlayerCount ^ 0.8 + 1 + tierWar / 4) / 6;           // Normalized to 1 == 5 players @ war tier 6
-    A3A_balancePlayerScale = A3A_balancePlayerScale * A3A_enemyBalanceMul;
+    A3A_balancePlayerScale = A3A_balancePlayerScale * (A3A_enemyBalanceMul / 10);
```
feel like there changes are out side the scope of the pr. Do they need to be merged in now for a bugfix? Or just minor changes?

jaj22 replied:
Within scope because it was a pain making fractions work with the UI, which is probably why the parameters screen didn't support it in the first place. Had a choice between making the UI even less readable or adding these. Switched a few times.

Was simpler to do the div 10 with initParams because it only required a one-way translation.

killerswin2 replied:
cool

killerswin2 commented on A3A/addons/core/functions/ModsAndDLC/fn_getModOfConfigClass.sqf:
```diff
@@ -21,7 +21,7 @@ if (count _addons > 0) then {
 
 if (_return == "") then {
-	_return = toLower getText (_config >> "DLC");
+	_return = tolower getText (_config >> "DLC");
```
you lowercase the command...... toLower

jaj22 replied:
Removed, discovered why it was there (BIS can't even keep their own DLC casing consistent), added it back in with different case :P

killerswin2 reviewed (changes requested):
Part 1 of review

jaj22 commented on A3A/addons/GUI/functions/SetupGUI/fn_setupDialog.sqf:
```diff
@@ -0,0 +1,161 @@
+    private _text = ["yr", "mon", "day", "hr", "min"];
+    private _range = [0, 12, 30, 24, 60];		// lazy but close enough
+
+    private _diffTime = [0, 0, 0, 0, 0];
+    for "_i" from 4 to 0 step -1 do {
+        _diffTime set [_i, _diffTime#_i + _newTime#_i - _oldTime#_i];
+        if (_i == 0 or _diffTime#_i >= 0) then { continue };
+        _diffTime set [_i, _diffTime#_i + _range#_i];
+        _diffTime set [_i-1, -1];
+    };
```
I don't see how. At best you'd need a forEach for the index, at which point it's saving exactly one line of code by using a BIS function.

killerswin2 replied:
yeah, we just need better scripting cmds

jaj22 commented on A3A/addons/GUI/functions/SetupGUI/fn_setupHQPosDialog.sqf:
```diff
@@ -0,0 +1,76 @@
+        private _posClicked = _mapCtrl posScreenToWorld [_xPos, _yPos];
+        private _posInvalid = true;
+
+        private _nearMarker = [markersX - controlsX, _posClicked] call BIS_fnc_nearestPosition;
```
Not seeing anything beneficial there. The visiblemap change isn't relevant here because the "current" HQ location is always valid.

killerswin2 replied:
ah, then it is time to kill #1764

jaj22 commented on A3A/addons/GUI/functions/SetupGUI/fn_setupHQPosDialog.sqf:
```diff
@@ -0,0 +1,76 @@
+            ["HQ Position", "Selected position cannot be in water."] call A3A_fnc_customHint;
+        };
+
+        if (_posClicked findIf { (_x < 0) || (_x > worldSize) } != -1) then {
```
Missed an exitWith here though. Not tested because you can't hit the case on either Altis or Malden due to the water check triggering first.

jaj22 commented on A3A/addons/GUI/functions/SetupGUI/fn_setupLoadgameTab.sqf:
```diff
@@ -0,0 +1,267 @@
+params["_mode", "_params"];
+
+Debug_1("Loadgame dialog called with mode %1", _mode);
+
```
There's already one in onLoad in setupDialog.sqf. The only way you'd call the tab functions without triggering that first is by external calls, which really aren't supported.

jaj22 commented on A3A/addons/GUI/functions/SetupGUI/fn_setupDialog.sqf:
```diff
@@ -0,0 +1,161 @@
+            ,["tem_anizay", "Anizay"]
+        ];
+        {
+            _x set ["mapStr", _prettyMapHM getOrDefault [_x get "map", _x get "map"]];
```
I thought that line was more readable than the next one, but whatever :P

jaj22 commented on A3A/addons/GUI/functions/SetupGUI/fn_setupDialog.sqf:
```diff
@@ -0,0 +1,161 @@
+            _ctrl ctrlShow (_x == _selectedTabIDC);
+        } forEach _allTabs;
+
+        switch (_selectedTab) do
```
Turns out `ctrlShow true` toggles all the child controls to true, so the update call needs to come after the ctrlShow loop. I shortened the function a bit though.

jaj22 commented on A3A/addons/GUI/dialogues/setupDialog.hpp:
```diff
@@ -0,0 +1,586 @@
+        class LoadgameTab : A3A_DefaultControlsGroup
+        {
+            idc = A3A_IDC_SETUP_LOADGAMETAB;
+            onLoad = "['onLoad'] spawn A3A_fnc_setupLoadgameTab";
```
Yep that worked.

Bob-Murphy commented:
Tried creating a new save. Vanilla only, two different CSAT as enemy. This popped up.
![image](https://user-images.githubusercontent.com/50651578/195213546-89ee6881-29fc-41b5-a31f-c9de950e8791.png)

jaj22 commented:
Missed a bunch of vanilla factions that I didn't know existed. Not entirely sure what they are. Would be helpful if something can sanity check:

CSAT_Enoch: Temperate camo, should force Contact DLC enabled.
CSAT_Apex: Tropical camo, should force Apex DLC enabled.
NATO_Apex: Tropical camo, should force Apex DLC enabled.

jaj22 commented:
Ok, I fixed up the template file for the new vanilla factions (turns out that they were added since I forked the setup UI stuff). However the forceDLC attributes are guesswork, as I don't know whether any DLC-locked gear or vehicles is actually present in these factions.

Bob-Murphy commented on A3A/addons/GUI/functions/SetupGUI/fn_setupDialog.sqf:
```diff
@@ -0,0 +1,154 @@
+            ,["chernarus_winter", "Chernarus (W)"]
+            ,["Enoch", "Livonia"]
+            ,["tem_anizay", "Anizay"]
+        ];
```
add the following for new maps in map overhaul:
,["chernarus_autumn", "Chernarus (A)"]
,["vn_khe_sanh", "Khe Sanh"]

Bob-Murphy commented on A3A/addons/GUI/functions/SetupGUI/fn_setupLoadgameTab.sqf:
```diff
@@ -0,0 +1,267 @@
+        if (_index == -1) exitWith {};
+
+        private _saveData = A3A_setup_saveData select _index;
+        private _str = format ["Really delete game with ID %1 on %2?", _saveData get "gameID", _saveData get "mapStr"];
```
stringify

Bob-Murphy commented on A3A/addons/GUI/functions/SetupGUI/fn_setupFactionsTab.sqf:
```diff
@@ -0,0 +1,254 @@
+        } forEach _savedFactions;
+
+        if (_failedFactions isNotEqualTo []) then {
+            private _msg = "Couldn't load factions from save:";
```
stringify?

Bob-Murphy commented on A3A/addons/GUI/functions/SetupGUI/fn_setupHQPosDialog.sqf:
```diff
@@ -0,0 +1,84 @@
+
+        private _nearMarker = [_display getVariable "dangerZones", _posClicked] call BIS_fnc_nearestPosition;
+        if (markerPos _nearMarker distance2d _posClicked < 500) exitWith {
+            ["HQ Position", "Selected position cannot be near enemy zones."] call A3A_fnc_customHint;
```
stringify?

Bob-Murphy commented on A3A/addons/GUI/functions/SetupGUI/fn_setupHQPosDialog.sqf:
```diff
@@ -0,0 +1,84 @@
+        };
+
+        if (surfaceIsWater _posClicked) exitWith {
+            ["HQ Position", "Selected position cannot be in water."] call A3A_fnc_customHint;
```
stringify?

jaj22 replied:
I'll do the HQ position and server information ones, but not the confirm dialog text, because it's really bad and needs a rewrite anyway. Adding stringtable functionality there will just make it harder to replace.

Bob-Murphy commented on A3A/addons/GUI/functions/SetupGUI/fn_setupHQPosDialog.sqf:
```diff
@@ -0,0 +1,84 @@
+        };
+
+        if (_posClicked findIf { (_x < 0) || (_x > worldSize) } != -1) exitWith {
+            ["HQ Position", "Selected position cannot be outside the map."] call A3A_fnc_customHint;
```
stringify?

Bob-Murphy commented on A3A/addons/GUI/functions/SetupGUI/fn_setupLoadgameTab.sqf:
```diff
@@ -0,0 +1,267 @@
+
+        private _occName = getText (A3A_SETUP_CONFIGFILE/"A3A"/"Templates"/_factionData#0#0/"name");
+        private _invName = getText (A3A_SETUP_CONFIGFILE/"A3A"/"Templates"/_factionData#0#1/"name");
+        _confirmText = _confirmText + endl + format ["%1 occupants and %2 invaders?", _occName, _invName];
```
stringify?

Bob-Murphy commented on A3A/addons/GUI/functions/SetupGUI/fn_setupLoadgameTab.sqf:
```diff
@@ -0,0 +1,267 @@
+            } else {
+                _saveData set ["startType", "load"];
+                _saveData set ["name", _oldSave getOrDefault ["name", ""]];
+                _confirmText = format ["Load game with ID %1, ", _oldSave get "gameID"];
```
stringify?

Bob-Murphy commented on A3A/addons/GUI/functions/SetupGUI/fn_setupLoadgameTab.sqf:
```diff
@@ -0,0 +1,267 @@
+            if (cbChecked _copyGameCtrl) then {
+                _saveData set ["startType", "copy"];
+                _saveData set ["name", ctrlText (_display displayCtrl A3A_IDC_SETUP_NAMEEDITBOX)];
+                _confirmText = format ["Copy game with ID %1, ", _oldSave get "gameID"];
```
stringify?

Bob-Murphy commented on A3A/addons/GUI/functions/SetupGUI/fn_setupLoadgameTab.sqf:
```diff
@@ -0,0 +1,267 @@
+            _saveData set ["startType", "new"];
+            _saveData set ["name", ctrlText (_display displayCtrl A3A_IDC_SETUP_NAMEEDITBOX)];
+            _saveData set ["startPos", markerPos "Synd_HQ"];
+            _confirmText = "Create new game with ";
```
stringify?

Bob-Murphy commented on A3A/addons/core/functions/Save/fn_saveLoop.sqf:
```diff
@@ -1,47 +1,71 @@
 savingServer = true;
 Info("Starting persistent save");
-["Persistent Save","Starting persistent save..."] remoteExec ["A3A_fnc_customHint",0,false];
+["Persistent Save","Starting persistent save..."] remoteExecCall ["A3A_fnc_customHint",0,false];
```
stringify?

Bob-Murphy commented on A3A/addons/core/functions/init/fn_startGame.sqf:
```diff
@@ -0,0 +1,22 @@
+// Throw message if background init isn't done
+if (isNil "A3A_backgroundInitDone") then {
+    Info("Waiting for background init to complete");
+    ["Server Information", "Waiting for background init to complete"] remoteExec ["A3A_fnc_customHint", 0];
```
stringify?

Bob-Murphy reviewed (changes requested):
I think I got all missing strings for the GUI and important information.
There are quite a few more customHints in some files you touched but that would be out ot scope.

HakonRydland commented on A3A/addons/core/Templates/AddonVics.hpp:
```diff
@@ -14,8 +14,8 @@ class AddonVics
     {
-        path = QPATHTOFOLDER(Templates\AddonVics);
-        requiredAddons[] = {"ToDo: Find patches entry"};
+        path = QPATHTOFOLDER(Templates\AddonVics\);
```
the `\` at the end should  really be moved to processing, but it can wait til a later date

jaj22 replied:
Nah, I'll probably do it. At one point it saved some messy string processing but now it's just two short lines that get slightly longer.

HakonRydland commented on A3A/addons/core/Templates/Templates.hpp:
```diff
@@ -1,533 +1,661 @@
+    {
+        name = "A3 CSAT Temperate";
+        file = "Vanilla_AI_CSAT_Enoch";
+        camo[] = {"temperate","tropical"};
```
ok so camo is actually suited for climates xyz...

jaj22 replied:
Can change to climate[]. Makes more sense given that it's not necessarily the actual camo for the troops.

HakonRydland reviewed (approved):
alright from a quick read and testing im ok with merger, while i have some disagreement about design choices for the underlying data and how its handled its simpler and functions perfectly fine... if any changes are wanted by the team it can be done later, as souch this is good to merge

jaj22 commented:
Changes:
- Added old fix for missing rank on save when dead.
- Fixed ammobox inventory undercover removal, probably.
- Moved joining backslashes in base paths to code.
- Changed camo[] to climate[] in templates.
- Added localization for HQ position dialog hints.
- Added localization for server information hints.
- Added new maps to name translation.

Things to do:
- Fix bug where params aren't always reset when they should be.
- Dig up and merge Bob's map texture changes (they move to a different file here, better to merge now).
- More testing.

Bob-Murphy commented:
init changes for map textures - already merged into unstable
https://github.com/official-antistasi-community/A3-Antistasi/pull/2452/files#diff-772a7d9b258c3c78599acd6dfe21a1d1d175ee3c5d182d0d72494e9753eb8081

jaj22 commented:
Turns out CUP still had large quantities of non-rifles in the rifle category, so I sorted that. Probably.

jaj22 commented:
Faction testing list: 
- ~~Check that all CUP factions load without errors.~~
- ~~Check that all vanilla factions load without errors.~~
- ~~Check that all GM factions load without errors.~~
- Check that all Unsung factions load without errors.
- Check that all VN factions load without errors.
- Check that all RHS factions load without errors.
- Check that all 3CBF factions load without errors.
- Check that all 3CB BAF factions load without errors.

Bob-Murphy reviewed (approved):
lgtm
